### ISSUE TITLE
`pulumi-eks` always download the latest `kubernetes` plugin 

### ISSUE THREAD

**Issue by aureq** [pulumi/pulumi-eks#800]
### What happened?

([Internal thread](https://pulumi.slack.com/archives/C0435R1H5FC/p1666677501385389?thread_ts=1666675408.522739&cid=C0435R1H5FC))

In Python, I have the following `requirements.txt` with all versions pinned, `pulumi-eks` seems to be dragging the latest versions of the `aws` and `kubernetes` plugins.

```
pulumi-aws==5.16.0
pulumi-eks==0.40.0
pulumi-kubernetes==3.21.4
```

After running `pip3 install -r requirements.txt`, I then run `pulumi plugin ls`.

```
❯ pulumi plugin ls
NAME        KIND      VERSION  SIZE    INSTALLED       LAST USED
aws         resource  5.18.0   427 MB  38 seconds ago  38 seconds ago
aws         resource  5.16.0   419 MB  1 minute ago    1 minute ago
eks         resource  0.40.0   200 MB  36 seconds ago  48 seconds ago
kubernetes  resource  3.22.0   84 MB   43 seconds ago  43 seconds ago
kubernetes  resource  3.21.4   83 MB   1 minute ago    1 minute ago

TOTAL plugin cache size: 1.2 GB
```

As shown above, both `aws` and `kubernetes` have 2 versions of their plugin installed.

If I run the same experiment but with `pulumi-eks==0.40.0` commented, running `pulumi plugin ls` shows

```
❯ pulumi plugin ls
NAME        KIND      VERSION  SIZE    INSTALLED       LAST USED
aws         resource  5.16.0   419 MB  6 seconds ago   6 seconds ago
kubernetes  resource  3.21.4   83 MB   30 seconds ago  30 seconds ago

TOTAL plugin cache size: 502 MB
```

Furthermore, if I delete the more recent plugins using `pulumi plugin rm resource kubernetes 3.22.0 --non-interactive --yes`, and the running `pulumi preview`. It appears the CLI download again the removed plugin.

What I've checked with the customer:
* they don't have any references to kubernetes 3.22.0 in their state file, only 3.21.4
* all their `requirements.txt` use pinned versions

### Steps to reproduce

1. pulumi plugin rm --all --yes
2. pulumi plugin ls
3. deactivate; 
4. rm -rf venv/;
5. python -m venv venv; 
6. source /home/aureq/work/customers/zendesk/1902/venv/bin/activate
7. pip3 install -r requirements.txt
8. pulumi plugin ls

### Expected Behavior

1. I should only see the following pulumi plugins installed that would match the pinned versions in `requirements.txt`.

    ```
    ❯ pulumi plugin ls
	NAME        KIND      VERSION  SIZE    INSTALLED       LAST USED
	aws         resource  5.16.0   419 MB  39 seconds ago  31 seconds ago
	eks         resource  0.40.0   200 MB  4 seconds ago   16 seconds ago
	kubernetes  resource  3.21.4   83 MB   1 minute ago    41 seconds ago
	```

2. When running `pulumi preview`, the list of plugins remain unchanged.

### Actual Behavior

1. Multiple pulumi plugins are installed and this prevents user/customers from pinning to specific versions when `pulumi-eks` is part of their Pulumi app.

2. Running `pulumi plugin rm resource kubernetes 3.22.0 --non-interactive --yes` and then `pulumi preview` will automatically install the removed plugin.

### Output of `pulumi about`

```
CLI          
Version      3.43.1
Go Version   go1.19.1
Go Compiler  gc

Plugins
NAME        VERSION
aws         5.16.0
eks         0.40.0
kubernetes  3.21.4
python      unknown

Host     
OS       debian
Version  11.5
Arch     x86_64

This project is written in python: executable='/home/aureq/work/customers/zendesk/1902/venv/bin/python3' version='3.9.14
'

Current Stack: case-1902

Found no resources associated with case-1902

Found no pending operations associated with case-1902

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/aureq
User           aureq
Organizations  aureq, team-ce, menfin, demo, pulumi

Dependencies:
NAME        VERSION
pip         22.0.4
pulumi-eks  0.40.0
setuptools  58.1.0
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Sefiris** [comment]
👍

Seeing the exact same issue in our projects right now.
Our requirements.txt looks like.
```
pulumi>=3.0.0,<4.0.0
pulumi-aws==5.10.0
pulumi-kubernetes==3.21.4
pulumi-eks==0.41.1
pulumi-mysql>=3.0.0,<4.0.0
```
And yet our pulumi previews still have a provider version update in them.

```
eks:index:Cluster$pulumi:providers:kubernetes (alpha-v1-alpha-01-eks-eks-k8s)
~  pulumi:providers:kubernetes (update)
    [id=589b1f59-92cc-4caf-b50f-bde7ed72d14e]
    [urn=urn:pulumi:dev::alpha-v1::eks:index:Cluster$pulumi:providers:kubernetes::alpha-v1-alpha-01-eks-eks-k8s]
    version   : "3.21.4" => "3.22.0"
 
eks:index:Cluster$pulumi:providers:kubernetes (alpha-v1-alpha-01-eks-provider)
~  pulumi:providers:kubernetes (update)
    [id=9788bc98-0a47-4f44-bda9-77fc207f1ac6]
    [urn=urn:pulumi:dev::alpha-v1::eks:index:Cluster$pulumi:providers:kubernetes::alpha-v1-alpha-01-eks-provider]
    version   : "3.21.4" => "3.22.0"
```

Normally i would not mind this, but currently i think this is a breaking change because this provider update is also causing a bunch of other replacements to be performed, some of them seem breaking and i do not feel comfortable performing this upgrade.

```
...
~   pulumi:providers:kubernetes   alpha-v1-alpha-01-eks-eks-k8s                                  updated   [diff: ~version]
~   pulumi:providers:kubernetes   alpha-v1-alpha-01-eks-provider                                 updated   [diff: ~version]
++  kubernetes:core:ConfigMap     alpha-v1-alpha-01-eks-nodeAccess                               replaced  
++  aws:ec2:SecurityGroupRule     alpha-v1-alpha-01-eks-tgw-kube-api-access                      replaced  
++  aws:iam:RolePolicyAttachment  alpha-v1-alpha-01-eks-policy-attach-vpc-resource-controller    replaced  
++  aws:ec2:SecurityGroupRule     alpha-v1-alpha-01-eks-default-pod-secgrp-linkerd-ingress-8086  replaced  
++  aws:ec2:SecurityGroupRule     alpha-v1-alpha-01-eks-default-pod-secgrp-dns-ingress           replaced  
++  aws:ec2:SecurityGroupRule     alpha-v1-alpha-01-eks-default-pod-secgrp-linkerd-ingress-8080  replaced  
++  aws:ec2:SecurityGroupRule     alpha-v1-alpha-01-eks-default-pod-secgrp-linkerd-ingress-8090  replaced  
++  aws:eks:NodeGroup             alpha-v1-alpha-01-eks-system3                                  replaced  
++  aws:eks:NodeGroup             alpha-v1-alpha-01-eks-pod-sec-apps                             replaced  
++  aws:eks:NodeGroup             alpha-v1-alpha-01-eks-internal-apps                            replaced  
...
```

**Comment by ringods** [comment]
It is not good that the latest plugin binary is retrieved and we will look further into this. Version 3.22.0 has Server Side Apply (SSA) on by default which is the reason for the suggested changes. Can you verify if disabling SSA is a good workaround for you?

```
$ pulumi config set kubernetes:enableServerSideApply false
```
to disable server side apply. If that doesn't work, do you mind setting an environment variable like this:

```
$ export PULUMI_K8S_ENABLE_SERVER_SIDE_APPLY=false
```

**Comment by Sefiris** [comment]
@ringods thank you, Setting the environment variable during the preview run stops provider update from updating a bunch of other resources. The version update is still applied off course as you've expected.

```
~   pulumi:providers:kubernetes  alpha-v1-alpha-01-eks-eks-k8s                           updated   [diff: ~enableServerSideApply, version]
~   pulumi:providers:kubernetes  alpha-v1-alpha-01-eks-provider                          updated   [diff: ~enableServerSideApply, version]
```

I can live with this workaround for now until this is hopefully fixed in a future version.

**Comment by danielrbradley** [comment]
Hi @aureq it looks like it could possibly be related to existing state which is already using the new version of these plugins. During any pulumi operation, if there are references in the state file to provider versions which aren't installed, it will install them automatically.

Please could you inspect your state and check if the newer provider versions are already referenced in there?

**Comment by danielrbradley** [comment]
So, after some investigation, this is a side effect of this being a component. The component chooses the version of the underlying providers it's going to use (aws and kubernetes). When writing the provider we reference a specific version of the dependencies to build and test against.

There is however a second issue at play here – the provider we distribute doesn't include a lock file so then we end up downloading the latest version of each dependency rather than the version we built against.

The behaviour where EKS manages its own sub-dependency versions is intended behaviour so that we can manage upgrades internally. The behaviour where it's automatically getting updated was a drawback of the original design but will be solved when we move to pkg to create standalone provider binary (https://github.com/pulumi/pulumi-eks/issues/728).

Closing as both intended behaviour and a duplicate as this will be solved by https://github.com/pulumi/pulumi-eks/issues/728
